start on api & extension docs & geyser system property docs (#159)
GeyserMC/GeyserWiki: _docs/geyser/geyser-api.md

---
title: Geyser API
---

Geyser has an API to extend what is possible with Geyser, and to allow you to interact with Geyser in your own plugins, mods, or extensions.

### Where can I use the Geyser API?
You could use the Geyser API in:
- A plugin for Paper/Spigot, Velocity, Waterfall/BungeeCord, etc.
- A mod for Fabric or Forge
- A Geyser Extension

### Accessing the Geyser API
See [here](/geyser/getting-started-with-api) for how to include the Geyser API dependency in your project.

### Documentation

The Geyser API offers events to subscribe to, or information on whether a player is joining through Geyser, and gives you the ability to enhance what is possible with Geyser (i.e. register custom items).
(soon, blocks and entities too).
It can be used easily in Geyser Extensions, see [here](/other/extensions) for details on those.

**Quick overview:** <br>
<div class="alert alert-info" role="alert">

> 	Note: To see full, detailed documentation, see the linked sources with javadocs.

Reviewer: ```suggestion
    Note: To see full, detailed documentation, see the linked sources with javadocs.
```

</div>

> #### [GeyserAPI](https://github.com/GeyserMC/Geyser/blob/master/api/src/main/java/org/geysermc/geyser/api/GeyserApi.java):

Reviewer 2: ```suggestion
#### [GeyserApi](https://github.com/GeyserMC/Geyser/blob/master/api/src/main/java/org/geysermc/geyser/api/GeyserApi.java):
```

The GeyserApi interface serves as a central access point to various functionalities provided by the Geyser API, providing methods to e.g. interact with player connections.

> It extends the [BaseAPI](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/GeyserApiBase.java) interface, which provides basic information about individual players.

Reviewer 2: ```suggestion
It extends the [Base API](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/GeyserApiBase.java) interface, which provides basic information about individual players.
```

To access it, you simply type:
```java
GeyserApi.api();
```

After you got the instance, you have access to all the methods.<br>
Use the documentation in the API module to see (and get info about) every single method available.  
Most [API methods have a simple explanation](https://github.com/GeyserMC/Geyser/blob/master/api/src/main/java/org/geysermc/geyser/api/GeyserApi.java).
Since the Geyser API extends the Base API that is shared across Floodgate and Geyser, you can also use the [methods from the Base API](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/GeyserApiBase.java).

#### We'll highlight a few to get you started quickly:
`GeyserApi#isBedrockPlayer(UUID)`<br>
Used to check if the given UUID of an **online** player is a Bedrock player.

`GeyserApi#connectionByUuid(UUID)`<br>
Used to get the [Connection](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/connection/Connection.java) of an **online** player.<br>
This method will return null if the player is not a Bedrock player.

<div class="alert alert-info" role="alert">

> 	You don't need to wait until the Bedrock player is online to use the getPlayer and isBedrockPlayer methods.<br>

Reviewer: ```suggestion
    You don't need to wait until the Bedrock player is online to use the getPlayer and isBedrockPlayer methods.<br>
```

    You can even use them in the pre-login events.
</div>

`GeyserApi#sendForm(UUID, Form(Builder))`<br>
Used to send a form to the Bedrock player with the given UUID.<br>
Click [here](/floodgate/forms/) to get more information about Forms.

`GeyserApi#onlineConnectionsCount()`<br>
Used to get the amount of online Bedrock players.

> ### Short Overview of the API

Reviewer: More a general question; should we already prepare for the Floodgate merge?
Because a lot of the stuff mentioned below aren't part of the base api and I think that that difference isn't mentioned enough. And this page is mainly focused on the geyser variant of the api.

Author: The geyser-api.md page should only show the geyser api imo; the floodgate-api page is separate.
With the merge, it would probably make sense to replace the floodgate-api page with a base-api page, and tell people that the Geyser API extends the base one

---
title: Geyser API
---

> ### Geyser has an API to extend what is possible and allow to listen to Geyser events. It also allows access to [Cumulus](/floodgate/forms/)

Reviewer 2: This should not be a subheading. Would break into `Events` and `Forms` with relevant info linked below each subheading for each.

This page outlines how to include the dependency for the Geyser API, and how to use the API.

#### Where can I use the Geyser API?
You could use the Geyser API in:

> - a plugin for Spigot/Paper, Velocity, Sponge, BungeeCord, etc.
> - a mod for Fabric/Forge
> - a Geyser Extension

Reviewer 2: ```suggestion
- A plugin for Paper/Spigot, Velocity, Waterfall/BungeeCord, etc.
- A mod for Fabric or Forge
- A Geyser Extension
```

### Accessing the Geyser API
See [here](/geyser/using-geyser-or-floodgate-as-a-dependency) for how to include the Geyser API in your project.

### Documentation

> The Geyser API offers events to subscribe to, or information on whether a player is joining through Geyser, and gives you the ability to e.g. register custom items 

Reviewer 2: ```suggestion
The Geyser API offers events to subscribe to, or information on whether a player is joining through Geyser, and gives you the ability to enhance what is possible with Geyser (i.e. register custom items).
```

(soon, blocks and entities too).
It can be used easily in Geyser Extensions, see [here](/other/extensions) for details on those.

> Quick overview:

Reviewer 2: ```suggestion
**Quick overview:**
```

Note: To see full, detailed documentation, see the linked sources with javadocs.

#### [GeyserAPI](https://github.com/GeyserMC/Geyser/blob/master/api/src/main/java/org/geysermc/geyser/api/GeyserApi.java):
The GeyserApi interface serves as a central access point to various functionalities provided by the Geyser API, providing methods to e.g. interact with player connections.
It extends the [BaseAPI](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/GeyserApiBase.java) interface, which provides basic information about individual players.

> The class GeyserAPI is the base class of the API and you need to use it to access any part of the API.<br>

Reviewer 2: ```suggestion
The class GeyserApi is the base class of the API and you need to use it to access any part of the API.<br>
```

Since the Geyser API extends the Base API that is shared across Floodgate and Geyser, you can also use the [methods from the Base API](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/GeyserApiBase.java).

#### We'll highlight a few to get you started quickly:
`GeyserApi#isBedrockPlayer(UUID)`<br>
Used to check if the given UUID of an **online** player is a Bedrock player.

`GeyserApi#connectionByUuid(UUID)`<br>
Used to get the [Connection](https://github.com/GeyserMC/api/blob/master/base/src/main/java/org/geysermc/api/connection/Connection.java) of an **online** player.<br>
This method will return null if the player is not a Bedrock player.

**Note**: You don't need to wait until the Bedrock player is online to use the getPlayer and isBedrockPlayer methods.<br>
You can even use them in the pre-login events.

`GeyserApi#sendForm(UUID, Form(Builder))`<br>
Used to send a form to the Bedrock player with the given UUID.<br>
Click [here](/floodgate/forms/) to get more information about Forms.

`GeyserApi#onlineConnectionsCount()`<br>
Used to get the amount of online Bedrock players.

> #### [Events](https://github.com/GeyserMC/Geyser/tree/master/api/src/main/java/org/geysermc/geyser/api/event):

Reviewer 2: Events should be in their own page IMO.

}
```
If you wish to listen to events in a Spigot/Paper plugin or a Fabric mod, you need to register the Geyser Event Bus as a listener first.
Extensions do not need to do that - they are automatically registered, so a simple @Subscribe annotation is enough.

**Fabric mod example:**
```java
public class ExampleMod implements ModInitializer, EventRegistrar {
    public static final Logger LOGGER = LoggerFactory.getLogger("modid");
    
    @Override 
    public void onInitialize() {
        ServerLifecycleEvents.SERVER_STARTING.register((server) -> {
            GeyserApi.api().eventBus().register(this, ExampleMod.class); // register your mod & this class as a listener
        });
        
        LOGGER.info("Geyser is cool!");
    }
    
    // here an event, we subscribe as usual with the @Subscribe annotation
    @Subscribe 
    public void onGeyserPostInitializeEvent(GeyserPostInitializeEvent eventad {
        LOGGER.info("Geyser started!");
    }
}
```
Do note: We cannot directly register the event but in the mod initializer, since the Geyser API would not be loaded yet.
Therefore, we register it in the server starting event provided by the Fabric API.

> **Spigot/Paper plugin example:**

Reviewer 2: ```suggestion
**Paper/Spigot plugin example:**
```

1. In your plugin.yml, add the following lines:
```yaml
  depend: ["Geyser-Spigot"]
```
This ensures that your plugin loads after Geyser has, so the Geyser API would be available.

2. In your main class, implement the EventRegistrar interface and register the event bus in the onEnable method:
```java
public class ExamplePlugin extends JavaPlugin implements EventRegistrar {
    
    @Override
    public void onEnable(){
        getLogger().info("Registering Geyser event bus!");

>         GeyserApi.api().eventBus().register(this, ExamplePlugin.class); // register your plugin & this class as a listener

Reviewer 2: ```suggestion
        GeyserApi.api().eventBus().register(this, this); // register your plugin & this class as a listener
```

    }

    // here an event, we subscribe as usual with the @Subscribe annotation
    @Subscribe
    public void onGeyserPostInitializeEvent(GeyserPostInitializeEvent event) {
        getLogger().info("Geyser started!");
    }
}
```

> 3. In case the '@Subscribe' annotation didn't work, you can also manually subscribe your method to the event bus:

Reviewer 2: ```suggestion
3. If you want to provide your event with a consumer, rather than annotating it, you can also manually subscribe your method to the event bus:
```
